### PR TITLE
Exposing config to cache ReadDir() response

### PR DESCRIPTION
**Pull request by raj-prince**
- Exposing functionality to cache the ReadDir() response in the kernel. This will drastically improve the performance of 2nd listing.
- Also, created another flag KeepDirContentPageCache, by setting it false invalidate the previously cached listing response.

### CODE REVIEW

File: fuseops/ops.go [hunk 1]
```diff
@@ -500,6 +500,14 @@ type OpenDirOp struct {
 	// a later call to ReleaseDirHandle.
 	Handle    HandleID
 	OpContext OpContext
+
+	// CacheDirContentAsPageCache conveys the kernel to cache the response of next
+	// ReadDirOp as page cache. Once cached, listing on that directory will be
+	// served from the kernel until invalidated.
+	CacheDirContentAsPageCache bool
```
Comment (stapelberg) [inline fuseops/ops.go:507]: Please rename this field to `CacheDir` for consistency with the Linux kernel flag name (`FOPEN_CACHE_DIR`). That the page cache is used for caching is an implementation detail.

File: fuseops/ops.go [hunk 2]
```diff
@@ -500,6 +500,14 @@ type OpenDirOp struct {
...
+
+	// KeepDirContentPageCache used to invalidate the cached directory content.
+	KeepDirContentPageCache bool
```
Comment (stapelberg) [inline fuseops/ops.go:510]: Please rename to `KeepCache` for consistency with the Linux kernel flag name (`FOPEN_KEEP_CACHE`). Also, the kernel documents this field as “don't invalidate the data cache on open”, so I think a better description is “KeepCache instructs the kernel to not invalidate the data cache on open calls”

**Review by stapelberg — CHANGES_REQUESTED** [review]
Do you have an example for how to use this? Could any of the existing examples be adjusted to exercise this feature?

**Comment by raj-prince** [comment]
> Do you have an example for how to use this? Could any of the existing examples be adjusted to exercise this feature?

Added as part of `samples/caching_fs`.

File: samples/cachingfs/caching_fs_test.go [hunk 3]
```diff
@@ -682,6 +683,7 @@ func (t *PageCacheTest) TwoFileHandles_NoKeepCache() {
 
 func (t *PageCacheTest) TwoFileHandles_KeepCache() {
 	t.fs.SetKeepCache(true)
+	fmt.Print(t.Dir)
```
Comment (stapelberg) [inline samples/cachingfs/caching_fs_test.go:686]: Remove debugging left-over

File: fuseops/ops.go [hunk 4]
```diff
@@ -500,6 +500,14 @@ type OpenDirOp struct {
 	// a later call to ReleaseDirHandle.
 	Handle    HandleID
 	OpContext OpContext
+
+	// CacheDir conveys the kernel to cache the response of next
```
Comment (stapelberg) [inline fuseops/ops.go:504]: conveys **to** the kernel

File: samples/cachingfs/caching_fs.go [hunk 5]
```diff
@@ -48,6 +48,11 @@ const (
 // Each file responds to reads with random contents. SetKeepCache can be used
 // to control whether the response to OpenFileOp tells the kernel to keep the
 // file's data in the page cache or not.
+//
+// Each directory responds to readdir with random entries (different names).
+// SetCacheDir and  SetKeepDirCache can be used to control whether the response
```
Comment (stapelberg) [inline samples/cachingfs/caching_fs.go:53]: nit: remove extra space

**Review by stapelberg — APPROVED** [review]

**Review by stapelberg — APPROVED** [review]

**Pull request merged**